File: packages/swt3-ai-ts/src/adapters/langgraph.ts
/**
 * SWT3 AI Witness SDK -- LangGraph Adapter.
 *
 * Wraps a LangGraph CompiledGraph to witness every graph invocation.
 * Supports invoke() and stream() methods (sync and async).
 *
 * Usage:
 *   import { wrapLangGraph } from "@tenova/swt3-ai/adapters/langgraph";
 *   const witnessed = wrapLangGraph(compiledGraph, witness);
 *   const result = await witnessed.invoke({ messages: [["user", "Hello"]] });
 *
 * Duck-typed: works with any object that has an invoke() method.
 * No langgraph import required.
 *
 * Copyright (c) 2026 Tenable Nova LLC. Apache 2.0. Patent pending.
 */

import type { Witness } from "../witness.js";
import type { InferenceRecord } from "../types.js";
import { sha256Truncated } from "../fingerprint.js";

export interface LangGraphGraph {
  invoke(input: unknown, ...args: unknown[]): unknown;
  stream?(input: unknown, ...args: unknown[]): unknown;
  name?: string;
}

function resolveModelId(graph: LangGraphGraph, explicit?: string): string {
  if (explicit) return explicit;
  if (typeof process !== "undefined") {
    if (process.env.SWT3_MODEL_ID) return process.env.SWT3_MODEL_ID;
    if (process.env.LANGGRAPH_MODEL) return process.env.LANGGRAPH_MODEL;
  }
  if (graph.name) return graph.name;
  return "langgraph-agent";
}

function stringifyValue(val: unknown): string {
  if (val === null || val === undefined) return "";
  if (typeof val === "string") return val;
  try {
    return JSON.stringify(val);
  } catch {
    return String(val);
  }
}

function buildRecord(
  modelId: string,
  input: unknown,
  output: unknown,
  elapsedMs: number,
): InferenceRecord {
  return {
    modelId,
    modelHash: sha256Truncated(modelId),
    promptHash: sha256Truncated(stringifyValue(input)),
    responseHash: sha256Truncated(stringifyValue(output)),
    latencyMs: elapsedMs,
    inputTokens: 0,
    outputTokens: 0,
    guardrailsActive: 0,
    guardrailsRequired: 0,
    guardrailPassed: true,
    hasRefusal: false,
    provider: "langgraph",
    guardrailNames: [],
  };
}

export function wrapLangGraph<T extends LangGraphGraph>(
  graph: T,
  witness: Witness,
  modelId?: string,
): T {
  const mid = resolveModelId(graph, modelId);
  const wrapped = Object.create(graph) as T;

  // Wrap invoke() -- handles both sync and async
  const originalInvoke = graph.invoke.bind(graph);
  wrapped.invoke = (input: unknown, ...args: unknown[]): unknown => {
    const start = performance.now();
    const result = originalInvoke(input, ...args);

    const finish = (res: unknown): unknown => {
      const elapsed = Math.round(performance.now() - start);
      witness.record(buildRecord(mid, input, res, elapsed));
      return res;
    };

    if (result && typeof (result as any).then === "function") {
      return (result as Promise<unknown>).then(finish);
    }
    return finish(result);
  };

  // Wrap stream() if present -- returns async generator
  if (graph.stream) {
    const originalStream = graph.stream.bind(graph);
    wrapped.stream = (input: unknown, ...args: unknown[]): unknown => {
      const result = originalStream(input, ...args);

      // If it returns an async iterable, wrap it
      if (result && typeof (result as any)[Symbol.asyncIterator] === "function") {
        const start = performance.now();
        return (async function* () {
          let lastChunk: unknown = undefined;
          for await (const chunk of result as AsyncIterable<unknown>) {
            lastChunk = chunk;
            yield chunk;
          }
          const elapsed = Math.round(performance.now() - start);
          witness.record(buildRecord(mid, input, lastChunk, elapsed));
        })();
      }

      // If it returns a sync iterable, wrap it
      if (result && typeof (result as any)[Symbol.iterator] === "function") {
        const start = performance.now();
        return (function* () {
          let lastChunk: unknown = undefined;
          for (const chunk of result as Iterable<unknown>) {
            lastChunk = chunk;
            yield chunk;
          }
          const elapsed = Math.round(performance.now() - start);
          witness.record(buildRecord(mid, input, lastChunk, elapsed));
        })();
      }

      return result;
    };
  }

  return wrapped;
}

File: packages/swt3-ai/swt3_ai/adapters/langgraph.py
"""SWT3 AI Witness SDK -- LangGraph Adapter.

Wraps a LangGraph CompiledGraph to witness every graph invocation.
Supports invoke(), ainvoke(), stream(), and astream() methods.

Usage:

    from swt3_ai.adapters.langgraph import wrap_langgraph

    witnessed = wrap_langgraph(compiled_graph, witness)
    result = witnessed.invoke({"messages": [("user", "Hello")]})

Configuration:
    1. SWT3_DSN=https://axm_live_xxx@sovereign.tenova.io/MY_ENCLAVE
    2. SWT3_ENDPOINT + SWT3_API_KEY + SWT3_TENANT_ID
    3. wrap_langgraph(graph, witness=my_witness)

Duck-typed: works with any object that has an invoke() method.
No langgraph import required.

Copyright (c) 2026 Tenable Nova LLC. Apache 2.0. Patent pending.
"""

from __future__ import annotations

import json
import logging
import os
import time
from typing import Any, Optional, TYPE_CHECKING

from ..fingerprint import sha256_truncated
from ..types import InferenceRecord

if TYPE_CHECKING:
    from ..witness import Witness

logger = logging.getLogger("swt3_ai.langgraph")


def _resolve_model_id(graph: Any, explicit: Optional[str] = None) -> str:
    if explicit:
        return explicit
    env_id = os.environ.get("SWT3_MODEL_ID") or os.environ.get("LANGGRAPH_MODEL")
    if env_id:
        return env_id
    if hasattr(graph, "name") and graph.name:
        return str(graph.name)
    return "langgraph-agent"


def _stringify(val: Any) -> str:
    if val is None:
        return ""
    if isinstance(val, str):
        return val
    try:
        return json.dumps(val, default=str)
    except Exception:
        return str(val)


def _build_record(
    model_id: str,
    input_val: Any,
    output_val: Any,
    elapsed_ms: int,
) -> InferenceRecord:
    return InferenceRecord(
        model_id=model_id,
        model_hash=sha256_truncated(model_id),
        prompt_hash=sha256_truncated(_stringify(input_val)),
        response_hash=sha256_truncated(_stringify(output_val)),
        latency_ms=elapsed_ms,
        input_tokens=0,
        output_tokens=0,
        guardrails_active=0,
        guardrails_required=0,
        guardrail_passed=True,
        has_refusal=False,
        provider="langgraph",
        guardrail_names=[],
    )


def wrap_langgraph(
    graph: Any,
    witness: Optional["Witness"] = None,
    model_id: Optional[str] = None,
    **overrides: Any,
) -> Any:
    """Wrap a LangGraph CompiledGraph to witness every invocation.

    Args:
        graph: Any object with an invoke() method (CompiledGraph or compatible).
        witness: Explicit Witness instance. If None, auto-creates from env.
        model_id: Model identifier override.
        **overrides: Override config (clearing_level, agent_id, etc.)

    Returns:
        Proxy object that behaves identically to the original graph
        but mints a witness anchor on each invocation.
    """
    if witness is None:
        try:
            from ..witness import SWT3Witness
            dsn = os.environ.get("SWT3_DSN")
            if dsn:
                witness = SWT3Witness.from_dsn(dsn, **overrides)
            else:
                endpoint = os.environ.get("SWT3_ENDPOINT", "")
                api_key = os.environ.get("SWT3_API_KEY", "")
                tenant_id = os.environ.get("SWT3_TENANT_ID", "")
                if endpoint and api_key and tenant_id:
                    witness = SWT3Witness(
                        endpoint=endpoint,
                        api_key=api_key,
                        tenant_id=tenant_id,
                        **overrides,
                    )
        except Exception:
            pass

    if witness is None:
        logger.debug("No witness configured; LangGraph adapter is a no-op")
        return graph

    mid = _resolve_model_id(graph, model_id)

    class LangGraphProxy:
        """Transparent proxy that witnesses LangGraph invocations."""

        __slots__ = ("__target", "__witness")

        def __init__(self, target: Any, w: "Witness"):
            object.__setattr__(self, "_LangGraphProxy__target", target)
            object.__setattr__(self, "_LangGraphProxy__witness", w)

        def invoke(self, input_val: Any, *args: Any, **kwargs: Any) -> Any:
            start = time.perf_counter()
            result = self.__target.invoke(input_val, *args, **kwargs)
            elapsed = round((time.perf_counter() - start) * 1000)
            record = _build_record(mid, input_val, result, elapsed)
            self.__witness.record(record)
            return result

        async def ainvoke(self, input_val: Any, *args: Any, **kwargs: Any) -> Any:
            if not hasattr(self.__target, "ainvoke"):
                raise AttributeError("ainvoke")
            start = time.perf_counter()
            result = await self.__target.ainvoke(input_val, *args, **kwargs)
            elapsed = round((time.perf_counter() - start) * 1000)
            record = _build_record(mid, input_val, result, elapsed)
            self.__witness.record(record)
            return result

        def stream(self, input_val: Any, *args: Any, **kwargs: Any) -> Any:
            if not hasattr(self.__target, "stream"):
                raise AttributeError("stream")
            start = time.perf_counter()
            last_chunk = None
            for chunk in self.__target.stream(input_val, *args, **kwargs):
                last_chunk = chunk
                yield chunk
            elapsed = round((time.perf_counter() - start) * 1000)
            record = _build_record(mid, input_val, last_chunk, elapsed)
            self.__witness.record(record)

        async def astream(self, input_val: Any, *args: Any, **kwargs: Any) -> Any:
            if not hasattr(self.__target, "astream"):
                raise AttributeError("astream")
            start = time.perf_counter()
            last_chunk = None
            async for chunk in self.__target.astream(input_val, *args, **kwargs):
                last_chunk = chunk
                yield chunk
            elapsed = round((time.perf_counter() - start) * 1000)
            record = _build_record(mid, input_val, last_chunk, elapsed)
            self.__witness.record(record)

        def __getattr__(self, name: str) -> Any:
            if name.startswith("_"):
                raise AttributeError(name)
            return getattr(self.__target, name)

    return LangGraphProxy(graph, witness)
